pkg/agent/flannel/setup.go

```diff
@@ -148,7 +150,40 @@ func createFlannelConf(nodeConfig *config.Node) error {
 		logrus.Infof("Using custom flannel conf defined at %s", nodeConfig.FlannelConfFile)
 		return nil
 	}
-	confJSON := strings.ReplaceAll(flannelConf, "%CIDR%", nodeConfig.AgentConfig.ClusterCIDR.String())
+	netMode, err := findNetMode(nodeConfig.AgentConfig.ClusterCIDRs)
+	if err != nil {
+		logrus.Fatalf("Flannel error checking netMode: %v", err)
+		return err
+	}
+	if netMode == ipv4 || netMode == (ipv4+ipv6) {
+		ipv4Enabled = "true"
+	} else {
+		ipv4Enabled = "false"
+	}
+	confJSON := strings.ReplaceAll(flannelConf, "%IPV4_ENABLED%", ipv4Enabled)
+	if netMode == ipv4 {
+		confJSON = strings.ReplaceAll(confJSON, "%CIDR%", nodeConfig.AgentConfig.ClusterCIDR.String())
+		confJSON = strings.ReplaceAll(confJSON, "%DUALSTACK%", "false")
+		confJSON = strings.ReplaceAll(confJSON, "%CIDR_IPV6%", emptyIPv6Network)
+	} else if netMode == (ipv4 + ipv6) {
+		confJSON = strings.ReplaceAll(confJSON, "%CIDR%", nodeConfig.AgentConfig.ClusterCIDR.String())
+		confJSON = strings.ReplaceAll(confJSON, "%DUALSTACK%", "true")
+		for _, cidr := range nodeConfig.AgentConfig.ClusterCIDRs {
+			if utilsnet.IsIPv6(cidr.IP) {
+				// Only one ipv6 range available. This might change in future: https://github.com/kubernetes/enhancements/issues/2593
```

Comment:
What would happen if there were more than one ipv6 range? I wonder if it would be worth checking and if there is more than one fail with an error

Reply:
I tested it configuring more then one CIDR of IPv4 and IPv6. K3s gives error and does not even start the logs say that CIDR len should be at least two on dualstack configuration.
If IPv6 is disabled the CIDR config for IPv6 it's not used.